cvat-cli/src/cvat_cli/_internal/agent.py

```diff
@@ -0,0 +1,337 @@
+import cvat_sdk.datasets as cvatds
+import urllib3.exceptions
+from cvat_sdk import Client, models
+from cvat_sdk.auto_annotation.driver import (
+    _AnnotationMapper,
+    _DetectionFunctionContextImpl,
+    _LabelNameMapping,
+    _SpecNameMapping,
+)
+from cvat_sdk.exceptions import ApiException
+
+from .common import CriticalError, FunctionLoader
+
+_POLLING_INTERVAL = timedelta(seconds=60)
+
+_UPDATE_INTERVAL = timedelta(seconds=30)
+
+
+class _RecoverableExecutor:
+    # A wrapper around ProcessPoolExecutor that recreates the underlying
+    # executor when a worker crashes.
+    def __init__(self, initializer, initargs):
+        self._mp_context = multiprocessing.get_context("spawn")
+        self._initializer = initializer
+        self._initargs = initargs
+        self._executor = self._new_executor()
+
+    def _new_executor(self):
+        return concurrent.futures.ProcessPoolExecutor(
+            max_workers=1,
+            mp_context=self._mp_context,
+            initializer=self._initializer,
+            initargs=self._initargs,
+        )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._executor.shutdown()
+
+    def submit(self, func, /, *args, **kwargs):
+        return self._executor.submit(func, *args, **kwargs)
+
+    def result(self, future: concurrent.futures.Future):
+        try:
+            return future.result()
+        except concurrent.futures.BrokenExecutor:
+            self._executor.shutdown()
+            self._executor = self._new_executor()
+            raise
+
+
+_current_function: cvataa.DetectionFunction
+
+
+def _worker_init(function_loader: FunctionLoader):
+    global _current_function
+    _current_function = function_loader.load()
+
+
+def _worker_job_get_function_spec():
+    return _current_function.spec
+
+
+def _worker_job_detect(context, image):
+    return _current_function.detect(context, image)
+
+
+class _Agent:
+    def __init__(self, client: Client, executor: _RecoverableExecutor, function_id: int):
+        self._client = client
+        self._executor = executor
+        self._function_id = function_id
+        self._function_spec = self._executor.result(
+            self._executor.submit(_worker_job_get_function_spec)
+        )
+
+        _, response = self._client.api_client.call_api(
+            "/api/functions/{function_id}",
+            "GET",
+            path_params={"function_id": self._function_id},
+        )
+
+        remote_function = json.loads(response.data)
+
+        self._validate_function_compatibility(remote_function)
+
+        self._agent_id = secrets.token_hex(16)
+        self._client.logger.info("Agent starting with ID %r", self._agent_id)
+
+        self._cached_task_id = None
+
+    def _validate_function_compatibility(self, remote_function: dict) -> None:
+        function_id = remote_function["id"]
+
+        if remote_function["provider"] != "native":
```

Comment:
Consider adding a named constant for `native`.

Reply:
Yeah, good point. Added.